CI workflow to create Github Docker Package after Successfull tests on commit

Amexei opened the pull request:
### About PR

Create a CD Github Actions pipeline that will test code and on success build/publish Docker image to Github Packages.

---

#### Workflow explanation

##### "ci.yml" - docker package build of the task

1. Folder "app" has angular folder with Dockerfile and nginx.conf. They are used in ci.yml Github Actions process.
2. ci.yml starts off with installation of node, used "20.15.0" version, the one that loaded the project on my local environment. And npm module installations.
3. `name: Run tests` step runs `npm run coverage`. Later steps use `if: success()` meaning they will run only if previous steps were successfull.
4. `name: Build` builds angular project and places it inside `app/angular` directory so `app/angular/dist/angular-starter/`
5. Later instead of building from CMD we use `buildx` Github Actions workflow that is much more straighforward and is easier to read. The step pushes docker image to Github packages.
6. Next step `name: Calculate Steps Log Time` instead of recording time variable at start and end then calculating duration we use ready-made repo `RockyIsHere/actions-calculate-log-time` for ease.
7. Last step is on `if: success()` or `if: failure()` we sent an email with **Job Status** and **Duration of workflow**

##### "docker-compose.yaml" - using our created in previous step package locally and executing curl

1. The docker-compose.yaml file is pretty short. Using `docker compose up ` we create a container from `image: "ghcr.io/amexei/angular-react-starter:lastest"` our previously created image and map local 80 port to container's 80 port.
2. To do curl we need to `docker exec -it ❗container_id❗ /bin/sh` and execute `curl localhost:80`
3. To check if created container `nginx` process is running as non-root we need to in same shell use `ps aux` as confirm that all processes are ran as `nginx` user.

---

#### Github /settings/secrets/actions Environment Variables to setup:

-   **GHUB_REPOSITORY** : Had to manually define "amexei/angular-react-starter" as "github.repository" has Higher case "A" letter and Github packages force you to use lowercase ghcr link.
-   **GHUB_TOKEN** : Token used for buildx and "RockyIsHere/actions-calculate-log-time". Can be created on https://github.com/settings/tokens
-   **MAIL_FROM** : Used in sending emails. Personal email.
-   **MAIL_TO** : Used in sending emails. Personal email.
-   **SMTP_USERNAME** : Gmail login
-   **SMTP_PASSWORD** : Gmail App Paasword https://myaccount.google.com/apppasswords

---

#### Main Useful links and Notes:

-   **Time measurment** - using https://github.com/marketplace/actions/github-workflow-duration-calculator | It uses Github API to get the workflow's time execution from start till the current step.
-   Added `"browsers": "ChromeHeadless"` to angular.js file to run tests in headless mode
-   Followed this guide for buildx docker package publish - https://cicube.io/workflow-hub/docker-build-push-action/#how-to-push-docker-images-to-github-container-registry-with-github-actions
-   Example of sending mail (had python execution in mind but this one uses specific Github Actions repo making it more easier) - https://github.com/go-gitea/gitea/issues/23725#issuecomment-1691883600
-   Two NGINX guides - https://medium.com/@bidyutnayak10/running-angular-apps-with-nginx-in-docker-anywhere-a-simple-guide-3e951b22f86f and https://dev.to/oneofthedevs/docker-angular-nginx-37e4
-   NGINX non-root solution - https://medium.com/kocsistem/how-to-run-nginx-for-root-non-root-5ceb13db6d41

Amexei commented:
@kseniyashaydurova and @dmitry-mightydevops Hello again. 👋🏻 

For ease, made the video of me going through the code and executing a test CI commit pipeline with email notification received and tests/build passing - 👉🏻 https://www.youtube.com/watch?v=IBNifr302Dw

The code in the video is exactly similar except for a slight change in folder structure to better tailor it for the task's description.

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,135 @@
+        runs-on: ubuntu-latest
+
+        steps:
+            # Bonus: STEAP 1 - Record time of job forund here - https://stackoverflow.com/a/72175840
```
Why it is commented? Let's remove commented stuff?

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,135 @@
+            - name: Setup Node.js
+              uses: actions/setup-node@v4
+              with:
+                  node-version: 20.15.0
```
Let's dynamically get node version from package.json

Amexei replied:
Initially after seeing  package.lock did not see "engines" so tried to find it in package-lock.json:
```
jq -r '.packages[] | .engines?.node? // empty' package-lock.json | sort -V | tail -n 1 | sed 's/||.*//'
```

The previous command outputs `^18.19.1` as it finds recursively all engines.node in multiple packages of package-lock.json file and finds the highest one.

Then looking at package.json I see `"version": "18.2.1"` which is similar to the parsed node version value. So is the project's node version defined in "version" key or is it version of the project itself?

Shall I leave it as is or parse the `"version": "18.2.1"` value from package.json?

kseniyashaydurova replied:
I just meant that you can use `https://github.com/actions/setup-node` GH action as it was before, but instead of hardcoding node veersion feed a `package.json` file to it

Amexei replied:
Noted 👍🏻 

Changed a bit to use `package.json`:
```
jq -r '.version' package.json
```

And added `^` to the value `18.2.1` in `"version"` is not present so it is going to take closest higher version.

Amexei replied:
UPD @kseniyashaydurova, 

Please note that as per Dmitry's change requested, I've added required node version field to the `package.json`. See more here https://github.com/Amexei/angular-react-starter/pull/2#discussion_r1813287123

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,135 @@
+                  node-version: 20.15.0
+
+            - name: Install packages
+              run: npm install
```
Usually it is better to use `npm ci` - https://builtin.com/software-engineering-perspectives/npm-ci-vs-npm-install

Amexei replied:
I've read the article. Changed to `npm ci` 🫡

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,135 @@
+
+            - name: Install packages
+              run: npm install
+              working-directory: angular
```
Let's specify default `working-directory` for a whole job to DRY

Amexei replied:
Added this code block after `runs-on: ubuntu-latest`:
```
defaults:
    run:
        working-directory: angular
```

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,135 @@
+              working-directory: angular
+
+            - name: Run tests
+              # No need for "continue-on-error: false" as it is Github Actions default
```
Thanks for comments, it is clear, we can remove it :)

Amexei replied:
Removed comments 🫡

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,135 @@
+              working-directory: angular
+
+            - name: Build
+              if: success()
```
No need  for `success` logic here, basically it is default Github Actions behavior to continue steps on `success` and to stop at failed step on `failure`

It is true for all `if: success()` checks below

Amexei replied:
Noted. Removed excessive `if: success()` checks.

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,135 @@
+              if: success()
+              uses: docker/login-action@v3
+              with:
+                  registry: ghcr.io
```
Let's pass registry as `input` variable

Amexei replied:
Before had `on:` specified as `push:` so after adding `workflow_call` or `workdlow_dispatch` inpusts were not specified.

Later found that workflow_call is used for using the current workflow inside other workflows. Could create other yml files to use inside main `ci.yml` one but in tasks's description it was forbidden.

Using `workflow_dispatch` allows for manual trigger, but at this moment manual triggers work only for main GH repo's branch so I had to use `gh` cli as described here - https://stackoverflow.com/a/67840292

```
gh workflow run 'CI Pipeline' --ref ci-feature
```

Could you please tell if this counts as a correct solution?

kseniyashaydurova replied:
See my reply in https://github.com/Amexei/angular-react-starter/pull/2#discussion_r1808520713

Amexei replied:
Addressed this part. Please check https://github.com/Amexei/angular-react-starter/pull/2#discussion_r1808520713 for a more detailed explanation.

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,135 @@
+              with:
+                  registry: ghcr.io
+                  username: ${{ github.actor }}
+                  password: ${{ secrets.GHUB_TOKEN }}
```
Let's use Github Actions automatic variables?
https://docs.github.com/en/actions/security-for-github-actions/security-guides/automatic-token-authentication
![image](https://github.com/user-attachments/assets/03746fff-1ba7-4b7d-9bfa-1a9df02c224d)

Amexei replied:
Using `github.token` now. 

Changed repos' Package rights as per https://stackoverflow.com/a/74205537

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,135 @@
+                  # Can't use github.repository as it's in uppercase my name Amexei
+                  #
+                  # Have to manually attach package to repository so it is visible inside repo https://stackoverflow.com/questions/73485013/github-package-successfully-published-but-not-showing-up-in-packages-section
+                  tags: ghcr.io/${{ secrets.GHUB_REPOSITORY }}:lastest
```
Take a look at this module to prepare docker tags and try to use it instead - https://github.com/docker/metadata-action

Amexei replied:
Done:
```
 - name: Preparing Docker tags
    id: meta
    uses: docker/metadata-action@v4
    with:
        images: ${{ inputs.registry }}/${{ secrets.GHUB_REPOSITORY }}
```

Want to note that I am using `secrets.GHUB_REPOSITORY` as my repo URI has Uppercase letter and `docker/*` steps fail with it.

Also changed docker-cmpose.yaml to have ci-feature tag as `metadata-action` by default makes the tag of the branch name instead of latest if it is not main or master.

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,135 @@
+                  # Have to manually attach package to repository so it is visible inside repo https://stackoverflow.com/questions/73485013/github-package-successfully-published-but-not-showing-up-in-packages-section
+                  tags: ghcr.io/${{ secrets.GHUB_REPOSITORY }}:lastest
+
+            # Bonus: STEP 2 - Record time of job forund here - https://stackoverflow.com/a/72175840
```
LEt's remove redundant stuff

Amexei replied:
Removed 🫡

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,135 @@
+              id: duration
+              uses: RockyIsHere/actions-calculate-log-time@v0.2.4
+              env:
+                  GITHUB_TOKEN: ${{ secrets.GHUB_TOKEN }}
```
The same as above related to existing GH Actions vars

Amexei replied:
Using `github.token` now. 👍🏻

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,135 @@
+            # If failure of success - https://stackoverflow.com/a/58859404
+            # Got an exanole from - https://github.com/go-gitea/gitea/issues/23725#issuecomment-1691883600
+            # Requires APP_PASSWORD from gmail https://stackoverflow.com/a/70929145
+            - name: Send success email
```
LEt's combine 2 emails to a single step

Amexei replied:
Done 🫡 

Now using `if: always()` on email job with conditional `${{ job.status }}` on email's subject.

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,135 @@
+              if: success()
+              uses: dawidd6/action-send-mail@v3
+              with:
+                  server_address: smtp.gmail.com
```
Specify this as workflow input with default value - https://docs.github.com/en/enterprise-cloud@latest/actions/writing-workflows/workflow-syntax-for-github-actions#onworkflow_callinputs

Amexei replied:
Done 🫡 

Described in the Change request above - https://github.com/Amexei/angular-react-starter/pull/2#discussion_r1808484762

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,135 @@
+              uses: dawidd6/action-send-mail@v3
+              with:
+                  server_address: smtp.gmail.com
+                  server_port: 587
```
Specify this as workflow input with default value - https://docs.github.com/en/enterprise-cloud@latest/actions/writing-workflows/workflow-syntax-for-github-actions#onworkflow_callinputs

Amexei replied:
Done 🫡 

Described in the Change request above - https://github.com/Amexei/angular-react-starter/pull/2#discussion_r1808484762

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,135 @@
+              with:
+                  server_address: smtp.gmail.com
+                  server_port: 587
+                  username: ${{ secrets.SMTP_USERNAME }}
```
Pls specify all secrets with description also in workflow secrets - https://docs.github.com/en/enterprise-cloud@latest/actions/writing-workflows/workflow-syntax-for-github-actions#onworkflow_callsecrets

Amexei replied:
@kseniyashaydurova Please clarify if I have to specifically use `workflow_call`?

So I will use the `ci.yml` as a big function and for instance create `main.yml` that will work on push? Is this allowed?

kseniyashaydurova replied:
Yes, it is allowed and that would be ok if you do a `workflow_call` to have inputs and secrets both defined.

Amexei replied:
Completed this part 👍🏻 

At the start tried to make several jobs that would pass parameters and artifacts from build process into the next job but got a thought on my mind that it was not exactly our goal as we would add 20+ lines of code and erase only 1 line on code.

So I've rebuilt the workflow to have one `trigger.yml` with secrets and inputs that would be triggered by a push and it is calling from inside the `reusable-upload-package.yml` as a reusable workflow component.

kseniyashaydurova replied:
Cool that you thought about such stuff, like adding 20+ lines to avoid a 1 line of code! It is very good developer quality to always ask yourself questions whether the change worth it or not.

kseniyashaydurova commented on Dockerfile:
```diff
@@ -0,0 +1,39 @@
+
+
+# ⭐ Serving the angular build on nginx server
+FROM nginx:alpine
```
Cool! But let's try to use existing `nginx` docker image instead

Amexei replied:
Could you please clarify what you mean by `existing nginx docker image`?

The task states to use:
`Inside that forked repo pls create Dockerfile based on an alpine image utilizing Nginx for the webserver.`

Could you please expand on the change requested?

Amexei replied:
This is the official NGINX dockerhub link:

https://hub.docker.com/_/nginx

Exactly this nginx image is used in my Dockerfile.

kseniyashaydurova replied:
Ok, you right

kseniyashaydurova commented on angular/angular.json:
```diff
@@ -79,39 +75,30 @@
-              "zone.js",
-              "zone.js/testing"
-            ],
+            "browsers": "ChromeHeadless",
```
Pls find a way to not change current `angular.json` file and pass required parameters from you GH action workflow.

This file shouldn't be changed

Amexei replied:
Found the syntax to passing parameters to a package.lock script with a `script -- [..args]`:

```
- name: Run tests
  run: npm run coverage -- --browsers=ChromeHeadless
```

Removed line in angualr.json file

kseniyashaydurova replied:
Pls then remove all your updates in angular.json (formatting stuff), like you didn't tiuch this file

Amexei replied:
Done 👍🏻 

It wasn't me it was Prettier 😸

kseniyashaydurova commented on docker-compose.yaml:
```diff
@@ -0,0 +1,15 @@
+        image: "ghcr.io/amexei/angular-react-starter:lastest"
+        ports:
+            - "80:80"
+        command: nginx -g "daemon off;"
```
Why is it needed?

Amexei replied:
If I understand correctly, the question is in relation to the `command: nginx -g "daemon off;"` line?

If yes, I've added this line so the docker container would not stop automatically on creation. 

Could you please clarify more in depth on this particular change requested?

Amexei replied:
UPD. Works without `command: nginx -g "daemon off;"`

Removed the line.

kseniyashaydurova replied:
Yes, I meant that you have it Dockerfile already

kseniyashaydurova commented on docker-compose.yaml:
```diff
@@ -0,0 +1,15 @@
+    web:
+        image: "ghcr.io/amexei/angular-react-starter:lastest"
+        ports:
+            - "80:80"
```
Ports Below 1024: On Unix-like systems, ports below 1024 are considered "privileged" ports.
So it's better to use something like 8080

Amexei replied:
Changed to:
```
web:
    image: "ghcr.io/amexei/angular-react-starter:lastest"
    ports:
        - "8080:80"
```

As I understand the comment is related to the local port of machine that is going to run this docker-compose.yaml file. So changed 80 to 8080. Now accessible on `curl localhost:8080`

kseniyashaydurova replied:
It is true for inner container too, cause it is also Linux OS, we try to never use such ports in our containers

Amexei replied:
Understood 👍🏻 

Changed `Dockerfile` and `nginx.conf` to use 8080 port in docker container.

Currently container's 8080 is mapped to our local 8080.

kseniyashaydurova commented on nginx.conf:
```diff
@@ -0,0 +1,22 @@
+  root /usr/share/nginx/html;
+
+  location / {
+    try_files $uri $uri/ /index.html =404;
```
this is unsafe solution
if your codebase contain sensitive files - you allow end users to render them, i.e. .env type of files.
check official docs what try_files does.

[nginx.org/en/docs/http/ngx_http_core_module.html#try_files](https://nginx.org/en/docs/http/ngx_http_core_module.html#try_files)

find a better and safer way for your particular case.

Amexei replied:
Changed to"
```
location / {
  index index.html;
}
```

Hard topic but found security related articles for the `try_files` and also mentions of it's speed slowdown compared to static paths. So changed to a static path.

kseniyashaydurova reviewed (changes requested):
Hi Aleksey, see my feedback:

1. I have problems with running it:
![image](https://github.com/user-attachments/assets/85b24662-fdfb-4b42-8a0d-76b4651456f2)

2. If I fix Dockerfile I have this, can you make you images registry public? So we would be able to check task functionality itself.
![image](https://github.com/user-attachments/assets/46732e63-0a0d-4a81-8990-1054fb0f2135)

3. Pls add screenshots with `success` and `failure` emails.

Amexei commented:
Hello again @kseniyashaydurova, 

Fixed all requested changes except for `Cool! But let's try to use existing nginx docker image instead` one as I do not fully understand what is required on this part. Waiting for clarifications 😀 

Regarding points in the comment:

1. `I have problems with running it:`

Removed `"name"` line, should work now. Please try on your side.

2.  `If I fix Dockerfile I have this, can you make you images registry public? So we would be able to check task functionality itself.`

Made package public.

3. `Pls add screenshots with success and failure emails.`

Success: 👇🏻 
![Success](https://media.discordapp.net/attachments/1242216344355868862/1297812488909361162/image.png?ex=67174991&is=6715f811&hm=cb0f54063d11ec29e8e6747196b0ee62073547c6c88528c76c0a8dcc09918a27&=&format=webp&quality=lossless)
Failure: 👇🏻 
![Fail](https://media.discordapp.net/attachments/1242216344355868862/1297812489224065085/image.png?ex=67174991&is=6715f811&hm=9c29f9ef23ba9e6d3c5e1378fe7850f5df051aedf0048addf6bb88bc4da6344d&=&format=webp&quality=lossless)

Amexei commented:
@kseniyashaydurova Sorry did not see the `10 hidden conversations' and asked for the change review preemptively.

**Will work on it now.** 

![Success](https://media.discordapp.net/attachments/1242216344355868862/1297827020201529374/image.png?ex=6717571a&is=6716059a&hm=96e3e9a7fb1db36604622a9d7aaa77ba991902445e7f655f70347a25cd8b3cbe&=&format=webp&quality=lossless&width=944&height=533)

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -1,128 +1,105 @@
+            registry:
+                description: Remote registry to push the Docker image to
+                default: ghcr.io
+                required: true
```
I think it shouldn't be required, because you defined `default`. Pls update it for all other inputs

Amexei replied:
Changed `true` to `false`.
Also removed `tag` input as using `docker/metadata-action` now.

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,112 @@
+            - name: Checkout code
+              uses: actions/checkout@v4
+
+            - name: Parse Node.js version from package-lock.json
```
See my comment here - https://github.com/Amexei/angular-react-starter/pull/2#discussion_r1809799410

Amexei replied:
Answered in https://github.com/Amexei/angular-react-starter/pull/2#discussion_r1809837805

It is not possible to parse version natively by `actions/checkout` as it does not contain `"version"` key inside `"node"` key.(i.e. node.version)

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,112 @@
+              env:
+                  GITHUB_TOKEN: ${{ github.token }}
+
+            - name: Print Calculated Time
```
LEt's remove this redundant step now?

Amexei replied:
Removed 👍🏻

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,112 @@
+name: CI Pipeline
+
+on:
```
Let's run this action on:
- push to `ci-feature` or `master` branch
- opening PR to `master` branch

Amexei replied:
Done 👍🏻. 

**Specified**:
```
on:
    push:
        branches:
            - ci-feature
            - master

    pull_request:
        branches:
            - master
        types: opened
```

kseniyashaydurova commented on Dockerfile:
```diff
@@ -0,0 +1,21 @@
+FROM nginx:alpine
+
+RUN rm -rf /usr/share/nginx/html/*
```
Can you pls check that all these operations from your Dockerfile are required?

Amexei replied:
This part was not useful at all as I was already removing the `default.conf` with my copy statement below.

kseniyashaydurova commented on Dockerfile:
Let's try to minimize image layers

Amexei replied:
Found a much shorter solution by using `nginxinc/nginx-unprivileged` and it is an official image https://hub.docker.com/r/nginxinc/nginx-unprivileged

Cleaned up `Dockerfile` and `nginx.conf`.

kseniyashaydurova commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,107 @@
+                  password: ${{ secrets.SMTP_PASSWORD }}
+                  subject: ${{ github.repository }} ${{ github.workflow }} ${{ job.status }}
+                  priority: high
+                  to: ${{ secrets.MAIL_TO }}
```
I think MAIL_TO and MAIL_FROM should be not secrets, just simple inputs

Amexei replied:
Done 👍🏻 

Added `mail_to` and `mail_from` as `required` inputs without default value.

kseniyashaydurova reviewed (changes requested):
Cool! Now it works! :fire:
![image](https://github.com/user-attachments/assets/475a516c-611d-42b5-b90b-24244fc54899)

But anyway left you some more feedback :slightly_smiling_face:

Pls `re-request` PR when ready

dmitry-mightydevops commented on .github/workflows/reusable-upload-package.yml:
```diff
@@ -0,0 +1,123 @@
+            - name: Checkout code
+              uses: actions/checkout@v4
+
+            - name: Parse Node.js version from package-lock.json
+              run: |
+                  VERSION=$(jq -r '.version' package.json)
+                  echo "VERSION=^$VERSION" >> $GITHUB_ENV
+
+            - name: Setup Node.js
+              uses: actions/setup-node@v4
+              with:
+                  node-version: ${{ env.VERSION }}
```
https://github.com/actions/setup-node
![image](https://github.com/user-attachments/assets/0b0fdffc-3dd7-45c4-808c-c2f0574f009b)

and adjust package.json as needed.

Amexei replied:
Doe 👍🏻 

Adjusted `package.json` with:
```
"engines": {
  "node": ">=18.19.1"
}
```

The reason for `>=18.19.1` is that this is the highest required version rom `package-lock.json` that I've parsed in https://github.com/Amexei/angular-react-starter/pull/2#discussion_r1808106268

dmitry-mightydevops commented on .github/workflows/reusable-upload-package.yml:
```diff
@@ -0,0 +1,123 @@
+                  push: true
+                  tags: ${{ steps.meta.outputs.tags }}
+
+            - name: Calculate Steps Log Time
+              if: always()
+              id: duration
+              uses: RockyIsHere/actions-calculate-log-time@v0.2.4
+              env:
+                  GITHUB_TOKEN: ${{ github.token }}
```
I never saw such action before and would consider it a **cheat**, but since you're the first to find and use it - I will accept it :)

I wanted you to show more "bashing" :)

Amexei replied:
Found it after spending some time bashing my head with bashing 🌚

dmitry-mightydevops commented on .github/workflows/trigger.yml:
```diff
@@ -0,0 +1,26 @@
+
+jobs:
+    commit-trigger:
+        uses: ./.github/workflows/reusable-upload-package.yml
```
let's give this file a better and more semantic name.

Amexei replied:
Changed `reusable-upload-package.yml` to `package-release.yml`

dmitry-mightydevops commented on .github/workflows/trigger.yml:
```diff
@@ -0,0 +1,26 @@
+            registry: ghcr.io
+            smtp_address: smtp.gmail.com
+            smtp_port: 587
+            mail_to: aleksei.melinic@gmail.com
+            mail_from: aleksei.melinic@gmail.com
```
so now your emails are exposed to the world! Welcome SPAM :)
let's convert actual email to secret.

Amexei replied:
Changed `mail_to` and 'mail_from` from inputs to secrets.

dmitry-mightydevops commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM nginxinc/nginx-unprivileged  
```
always Pin/TAG your images, because next latest may have a breaking changes 
your builds (if no data changes) should reproduce exactly the same, in your solution it will not.

immutable images are GOOD
This (immutable) enables immutable infrastructure. It guarantees that deployments are idempotent — they will stay exactly the same no matter how many times you repeat the operation.

Amexei replied:
You are right, good suggestion.

Added `:1.27` tag as it is the current `latest` one.

Amexei replied:
UPD. Changed `:1.27` to `:1.27-alpine` as I've found that regular debian image is less secure than alpine plus it was requested to be an `alpine` image in the task's description/

dmitry-mightydevops commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+
+EXPOSE 8080
+
+CMD ["nginx", "-g", "daemon off;"]
```
you dockerfile should have explicity USER prior to the CMD for readability
> explicit is better than implicit 

(a good python rule)

I know you have used nginx-unprivileged, but this is implicit.

Amexei replied:
Understood 🫡 

Added `USER nginx` command.

Amexei commented on .github/workflows/reusable-upload-package.yml:
```diff
@@ -0,0 +1,123 @@
+                required: true
+
+jobs:
+    test-build-publish:
```
Noted 👍🏻 

Changed to `test-build-publish` to `build`.

In `trigger.yml` changed job name `commit-trigger` to `push-trigger` which is ore semantic.

Amexei commented on .github/workflows/reusable-upload-package.yml:
```diff
@@ -0,0 +1,123 @@
+                description: Email address to send notifications from
+                required: true
+                type: string
+            ghub_repository:
+                description: GitHub Repository link with lowercase
+                default: amexei/angular-react-starter
+                required: false
+                type: string
```
I tried to use the `github.repository` but my github name has an uppercase letter giving => `Amexei/angular-react-starter` which gets rejected by Github Package Registry for having an uppercase letter.

I can add another step to get the `github.repository` and make sure it is `,,` in lowercase. Shall I do this or elave it as is?

Amexei commented on .github/workflows/trigger.yml:
Applied `yamlfix` for all `.yaml` files.

Amexei replied:
UPD.

Ksenia requested to have `newline` delimiter between steps here https://github.com/Amexei/angular-react-starter/pull/2#discussion_r1814152945

Added `yamlfix-config.toml` for `yamlfix` so these settings allow for white lines between steps and section:
```toml
whitelines = 1
section_whitelines = 1
```

Amexei commented on .github/workflows/reusable-upload-package.yml:
Done 👍🏻

Amexei replied:
UPD.

Ksenia requested to have `newline` delimiter between steps here https://github.com/Amexei/angular-react-starter/pull/2#discussion_r1814152945

Added `yamlfix-config.toml` for `yamlfix` so these settings allow for white lines between steps and section:
```toml
whitelines = 1
section_whitelines = 1
```

Amexei replied:
UPD.

Ksenia suggested using pre-commit library but for easy added in `.git` a `pre-commit` file with:

```
yamlfix -c yamlfix-config.toml ./.github/.
yamlfix -c yamlfix-config.toml ./docker-compose.yaml

git add .
```

So files get `yamlfixed` before each commit.

Amexei commented:
Dear @kseniyashaydurova and @dmitry-mightydevops changes requested are addressed, please review 🙇🏻‍♂️

kseniyashaydurova commented on .github/workflows/package-release.yml:
```diff
@@ -0,0 +1,101 @@
+      run:
+        working-directory: angular
+    steps:
+      - name: Checkout code
```
Let's separate each step with empty line to have better readability

Amexei replied:
Resolved this by adding a config `yamlfix-config.toml` for `yamlfix`.

kseniyashaydurova commented on angular/package.json:
```diff
@@ -41,5 +41,8 @@
     "typescript": "5.5.4",
     "typescript-eslint": "8.2.0"
+  },
+  "engines": {
```
Usually `engines` info is located at the top right after `name` and `version` fields, cause this information may be needed quite often and it is good to locate such stuff somewhere at the top fir easier and faster search

Amexei replied:
You are right, shifted it to the top.

kseniyashaydurova commented on nginx.conf:
```diff
@@ -0,0 +1,12 @@
+
```
Pls remove redundanr empty lines. Usually we don't add an empty line at the file start and have a single empty line at the end of the file

Amexei replied:
I really enjoy having good and consistent formatting, but the empty space at the end of file is added not for me but by https://lyz-code.github.io/yamlfix/ formatter that was requested to be used by Dmitry. 

Removing empty lines would break yaml format style of `yamlfix` as it has empty line in its rules:

![Features](https://media.discordapp.net/attachments/1242216344355868862/1298847388047507506/image.png?ex=671b0d64&is=6719bbe4&hm=41fe52277048fb5a97e6d0d5c30458be053095ebac306494937d1a63af3f8c7c&=&format=webp&quality=lossless&width=706&height=533))

Please see Dmitry's comment: https://github.com/Amexei/angular-react-starter/pull/2#discussion_r1813285241 

What shall I do?

Amexei replied:
UPD.

Found in the `yamlfix` documentation Reference to `adapters,py` https://lyz-code.github.io/yamlfix/reference/#yamlfix.adapters.SourceCodeFixer and changed in my `venv` function called `_add_newline_at_end_of_file` so it does not creates newline.

kseniyashaydurova replied:
We usually use automatic `git pre-commit` hooks for that to do it automatically on commit: https://pre-commit.com/

So we usually add similar to these lines in our `.pre-commit-config.yaml` - https://pre-commit.com/#2-add-a-pre-commit-configuration (i.e. `chack-yaml` to format yaml files in a nice way, `end-of-file-fixer` to leave only one empty line at the end of each file, `trailing-whitespace` to remove trailing whitespaces at the end of lines)

Amexei replied:
Noted,

Going to learn that.

Amexei replied:
UPD.

Added custom `pre-commit` hook to `.git` folder that runs `yamlfix` before commit.

kseniyashaydurova reviewed (approved):
Great! Now it works and looks nice and elegant! You can do few minor fixes to beautify it a bit more :)

kseniyashaydurova commented on nginx.conf:
```diff
@@ -0,0 +1,12 @@
+
```
Let's user for this nginx conf

Amexei replied:
Tried to change nginx.conf and Dockerfile but it did not work out as I am using `nginxinc/nginx-unprivileged` which does not allow this setting in .conf file(or at least requires running main process with root privileges).

Got the following warn:
```bash
nginx: [warn] the "user" directive makes sense only if the master process runs with super-user privileges, ignored in /etc/nginx/nginx.conf:1
```

So as I understand to run nginx with a `user:` command as requested I would have to start nginx main process as a root. The `nginxinc/nginx-unprivileged` image runs nginx as `nginx` user by default even without specifying it within Dockerfile.

Please see https://stackoverflow.com/a/72642450:

![image](https://media.discordapp.net/attachments/1242216344355868862/1298876517597843476/image.png?ex=671b2885&is=6719d705&hm=76668f1c63b7d5cf32fb7547c0cf6f32d9eb62c3d4606f72f49775613c4f2d91&=&format=webp&quality=lossless)

dmitry-mightydevops commented on Dockerfile:
```diff
@@ -0,0 +1,10 @@
+
+USER nginx
+
+CMD ["nginx", "-g", "daemon off;"]
```
### CMD should not contain config options

you should have a single source of truth - which is your nginx.conf
otherwise the maintenance will be problematic. You changed the config and your expect it to work, but it didn't because some config options were overridden somewhere else.

Move your -g daemon off to nginx.conf

dmitry-mightydevops commented on Dockerfile:
```diff
@@ -0,0 +1,10 @@
+FROM nginxinc/nginx-unprivileged:1.27-alpine
```
add link to 
- dockerfile or official github or dockerhub so we can quickly reference it and obtain documentation as necessary

simple maintenance in the future - is key to have SANE workflows with the team of multiple engineers.

Amexei replied:
Done.

Before I had `nginx.conf` copied into `default.conf` file so it was technically `default.conf`

Changed copying to `/etc/nginx/nginx.conf` and added `daemon off;` line.

Amexei replied:
Added link to specific alpine version on github. From there it is easy to navigate to the docs.

dmitry-mightydevops commented on yamlfix-config.toml:
all util type configs should have a dot prefix i.e. **.yamlfix.toml**

Amexei replied:
Renamed to `config.yamlfix.toml`

dmitry-mightydevops commented on node/swagger.yaml:
don't keep files you don't need.

Amexei replied:
Removed

dmitry-mightydevops commented on .github/workflows/package-release.yml:
```diff
@@ -0,0 +1,122 @@
+---
+name: CI Pipeline
+
+on:
```
Add doctring above your reusable workflow and explain how to use
provide samples

your goal is to make life easier for anyone who would be using your solution (or maintaining it)

here is an example:


```yaml
# --------------------------------------------------------------
# GitHub Actions Workflow: <Workflow Name>
# 
# Description:
#   This workflow automates the process of [describe the primary task 
#   or purpose, e.g., "building and testing the project," "deploying 
#   the application to production," etc.]. It consists of multiple 
#   jobs with various steps to ensure [mention any goals like "code 
#   quality," "successful deployment," or "artifact generation"].
#
# Triggers:
#   - on: 
#       - push: Executes on a push to [specific branches or paths if any].
#       - pull_request: Executes on a pull request to [specific branches].
#       - schedule: Runs on a cron schedule for [mention frequency, 
#         e.g., "daily builds" or "weekly maintenance"].
#
# Jobs:
#   - <Job Name 1>: [Brief description of the job, e.g., "Sets up 
#       the environment and installs dependencies."]
#   - <Job Name 2>: [e.g., "Runs unit and integration tests to 
#       validate changes."]
#   - <Job Name 3>: [e.g., "Deploys to the production environment 
#       upon successful completion of all tests."]
#
# Notes:
#   - Requires the following secrets to be configured in the GitHub 
#     repository: [list secrets, e.g., "AWS_ACCESS_KEY", "DOCKER_TOKEN"].
#   - [Any prerequisites or dependencies, e.g., "Ensure Docker is 
#     installed on self-hosted runners."]
#   - [Any environment configurations or versions required, e.g., 
#     "Node.js 14+" or "Python 3.9"]
#
# Expected Outcome:
#   - Successful execution of this workflow should result in [state 
#     the outcome, e.g., "a new Docker image published to Docker Hub" 
#     or "a deployed app on production servers"]. 
#
# --------------------------------------------------------------

```

Amexei replied:
Added description of my reusable workflow

dmitry-mightydevops commented on .github/workflows/trigger.yml:
```diff
@@ -0,0 +1,23 @@
+name: CI Pipeline
+
+on:
+  push:
+    branches: [ci-feature, master]
+  pull_request:
+    branches: [master]
+    types: opened
+
```
use list one per line (see how to do that with yamlfix configs) not [x, y, z] - it's harder to read or compare files with diff.

Amexei replied:
Added `sequence_style = "block_style"` to `yamlfix-config.toml`

dmitry-mightydevops commented on .github/workflows/trigger.yml:
```diff
@@ -0,0 +1,23 @@
+---
+name: CI Pipeline
```
prefer lowercases
```suggestion
name: build-pipeline
```

Amexei replied:
Done.

Renamed file to `build-pipeline.yml` and changed name to `build-pipeline`

dmitry-mightydevops commented on .github/workflows/package-release.yml:
```diff
@@ -0,0 +1,122 @@
+---
+name: CI Pipeline
```
```suggestion
name: build-workflow
```

and name file build-workflow.yml or name: package-release whatever you prefer more.

Amexei replied:
Renamed as `build-workflow.yml`

dmitry-mightydevops commented on .github/workflows/trigger.yml:
```diff
@@ -0,0 +1,23 @@
+  push-trigger:
+    uses: ./.github/workflows/package-release.yml
+    with:
+      registry: ghcr.io
+      smtp_address: smtp.gmail.com
+      smtp_port: 587
+      ghub_repository: amexei/angular-react-starter
+    secrets:
+      smtp_username: ${{ secrets.SMTP_USERNAME }}
+      smtp_password: ${{ secrets.SMTP_PASSWORD }}
+      mail_to: ${{ secrets.MAIL_TO }}
+      mail_from: ${{ secrets.MAIL_FROM }}
```
sort all keys by ASC

Amexei replied:
Sorted keys in Alphabetical order by ASC

dmitry-mightydevops commented on nginx.conf:
```diff
@@ -0,0 +1,52 @@
+    access_log  /var/log/nginx/access.log  main;
+
+    sendfile        on;
+    #tcp_nopush     on;
+
+    keepalive_timeout  65;
+
+    #gzip  on;
+
```
don't have commented code - they're always noise. If you want to keep it - then you should comment in what case to turn it on. 

this is valid for some complex solutions that we keep commenting, but not for the basic ones like tcp_nopush or gzip

Amexei replied:
Removed comments

dmitry-mightydevops commented on nginx.conf:
```diff
@@ -0,0 +1,52 @@
+
+daemon off;
+
+error_log  /var/log/nginx/error.log notice;
+pid        /tmp/nginx.pid;
+
+
+events {
+    worker_connections  1024;
+}
+
+
+http {
+    proxy_temp_path /tmp/proxy_temp;
+    client_body_temp_path /tmp/client_temp;
+    fastcgi_temp_path /tmp/fastcgi_temp;
+    uwsgi_temp_path /tmp/uwsgi_temp;
+    scgi_temp_path /tmp/scgi_temp;
```
your formatting should be consistent
if you formatted error_log, pid to have the same column for the value, it should be similar for other options like 
proxy_temp_path etc

Amexei replied:
Fixed indentations to follow one style in order to be more readable.
